lib/semantics/expression.cc

```diff
@@ -29,25 +29,53 @@
 
 using namespace Fortran::parser::literals;
 
+// Typedef for optional generic expressions (ubiquitous in this file)
+using MaybeExpr =
+    std::optional<Fortran::evaluate::Expr<Fortran::evaluate::SomeType>>;
+
 // Much of the code that implements semantic analysis of expressions is
 // tightly coupled with their typed representations in lib/evaluate,
 // and appears here in namespace Fortran::evaluate for convenience.
 namespace Fortran::evaluate {
 
 using common::TypeCategory;
 
-using MaybeExpr = std::optional<Expr<SomeType>>;
+// Constraint checking
+void ExpressionAnalysisContext::CheckConstraints(MaybeExpr &expr) {
+  if (inner_ != nullptr) {
+    inner_->CheckConstraints(expr);
+  }
+  if (constraint_ != nullptr && expr.has_value()) {
```

Comment:
This function might be clearer if it had `if (expr.has_value()) {...}` around the whole body. At first glance it seems like the first `if` might do something on a `nullopt` expr, but I don't think it does.

Reply:
Ok.

Reply:
This was a bad suggestion on my part. The call to `inner_->CheckConstraints(expr)` can cause expr to be reset, so we **do** need to check `expr.has_value()` before trying to use the contained value.